File: scripts/lib/blog-prompt.ts
import type { BlogSeriesConfig } from "./blog-series-config.ts";
import type { BlogPost } from "./blog-posts.ts";
import type { BlogComment } from "./blog-comments.ts";
import {
  TWEET_SECTION_HEADER,
  BLUESKY_SECTION_HEADER,
  MASTODON_SECTION_HEADER,
} from "./types.ts";

export interface BlogContext {
  readonly series: BlogSeriesConfig;
  readonly agentsMd: string;
  readonly previousPosts: readonly BlogPost[];
  readonly comments: readonly BlogComment[];
  readonly today: string;
}

const dayOfWeek = (dateStr: string): number => new Date(dateStr + "T12:00:00").getDay();
const isSunday = (dateStr: string): boolean => dayOfWeek(dateStr) === 0;

const RECAP_KEYWORDS = ["weekly recap", "monthly recap", "quarterly recap", "annual recap"] as const;

const isRecapPost = (post: BlogPost): boolean =>
  RECAP_KEYWORDS.some((keyword) => post.title.toLowerCase().includes(keyword));

const postsSinceLastRecap = (posts: readonly BlogPost[], today: string): readonly BlogPost[] => {
  const todayDow = dayOfWeek(today);
  const daysSinceSunday = todayDow === 0 ? 7 : todayDow;
  const lastRecapIdx = posts.findIndex(isRecapPost);
  const postsSinceRecap = lastRecapIdx >= 0 ? posts.slice(0, lastRecapIdx + 1) : posts.slice(0, daysSinceSunday + 1);
  return postsSinceRecap.slice(0, 7);
};

const EMBED_HEADERS = [TWEET_SECTION_HEADER, BLUESKY_SECTION_HEADER, MASTODON_SECTION_HEADER] as const;

export const stripEmbedSections = (body: string): string => {
  const firstEmbedIndex = EMBED_HEADERS
    .map((header) => body.indexOf(header))
    .filter((index) => index >= 0)
    .reduce((min, index) => Math.min(min, index), body.length);
  return body.slice(0, firstEmbedIndex).trimEnd();
};

const formatFullPost = (post: BlogPost): string =>
  `\n### ${post.title} (${post.date})\n${stripEmbedSections(post.body)}\n`;

const buildPostHistory = (posts: readonly BlogPost[], today: string): string => {
  if (posts.length === 0) return "";
  const relevant = postsSinceLastRecap(posts, today);
  const full = relevant.map(formatFullPost).join("");
  const older = posts.slice(relevant.length).map((p) => `- ${p.title} (${p.date})\n`).join("");
  return `\n\n## Previous Posts (most recent first)\n${full}${older ? `\n### Older posts (titles only)\n${older}` : ""}`;
};

const buildCommentsSection = (comments: readonly BlogComment[]): string => {
  if (comments.length === 0) return "\n\n📝 No reader comments yet — comments are very rare on this site, so if any appear, prioritize serving those readers.";
  const sorted = [...comments].sort((a, b) => a.isPriority === b.isPriority ? 0 : a.isPriority ? -1 : 1);
  const formatted = sorted.map((c) =>
    `\n**${c.author}**${c.isPriority ? " ⭐ PRIORITY" : ""} (${c.createdAt}):\n${c.body}\n`
  ).join("");
  return `\n\n## Reader Comments (from Giscus/GitHub Discussions)\n⭐ Comments are rare on this site — treat every commenter as a VIP and steer toward their interests.\n${formatted}`;
};

const recapInstructions = (today: string): string => {
  const date = new Date(today + "T12:00:00");
  const dow = date.getDay();
  const isLastDayOfMonth = new Date(date.getFullYear(), date.getMonth() + 1, 0).getDate() === date.getDate();
  const month = date.getMonth();
  const isLastDayOfQuarter = isLastDayOfMonth && [2, 5, 8, 11].includes(month);
  const isLastDayOfYear = month === 11 && date.getDate() === 31;

  if (isLastDayOfYear) return "\n\n🗓️ TODAY IS THE LAST DAY OF THE YEAR. Write an **Annual Recap** summarizing the quarterly recaps from this year. Title format: YYYY | 📊 Annual Recap";
  if (isLastDayOfQuarter) return "\n\n🗓️ TODAY IS THE LAST DAY OF THE QUARTER. Write a **Quarterly Recap** summarizing the monthly recaps from this quarter. Title format: YYYY-MM-DD | 📊 Quarterly Recap";
  if (isLastDayOfMonth) return "\n\n🗓️ TODAY IS THE LAST DAY OF THE MONTH. Write a **Monthly Recap** summarizing the weekly recaps from this month. Title format: YYYY-MM-DD | 📊 Monthly Recap";
  if (dow === 0) return "\n\n🗓️ TODAY IS SUNDAY. Write a **Weekly Recap** summarizing the past 6 days of posts. Title format: YYYY-MM-DD | 📊 Weekly Recap";
  return "";
};

export const buildBlogPrompt = (context: BlogContext): { system: string; user: string } => {
  const { series, agentsMd, previousPosts, comments, today } = context;

  const system = agentsMd || `You are ${series.name}, an automated blog. Write a blog post.`;
  const user = `Write a new blog post for today, ${today}.
${buildPostHistory(previousPosts, today)}${buildCommentsSection(comments)}${recapInstructions(today)}

## Instructions

Generate ONLY the blog post body in markdown (starting with a ## heading).
Do NOT generate frontmatter — it will be added automatically.
Do NOT wrap your output in code fences.

${previousPosts.length === 0
    ? "This is the FIRST post in the series. Introduce yourself and set the tone."
    : `Continue the series naturally.`}

${comments.length > 0
    ? "Address reader comments naturally in your post. Only incorporate what fits."
    : ""}`;

  return { system, user };
};

export const filterCommentsAfterLastPost = (
  comments: readonly BlogComment[],
  previousPosts: readonly BlogPost[],
  postTimeUtc: string,
): readonly BlogComment[] => {
  if (previousPosts.length === 0) return comments;
  const lastPostDate = previousPosts[0]!.date;
  const cutoff = `${lastPostDate}T${postTimeUtc}:00Z`;
  return comments.filter((c) => c.createdAt >= cutoff);
};

export const buildBackLink = (series: BlogSeriesConfig, previousPost: BlogPost): string =>
  `[[${series.id}/${previousPost.filename.replace(/\.md$/, "")}|⏮️]]`;

export const buildForwardLink = (series: BlogSeriesConfig, nextFilename: string): string =>
  `[[${series.id}/${nextFilename.replace(/\.md$/, "")}|⏭️]]`;

const quoteForYaml = (value: string): string =>
  `"${value.replace(/\\/g, "\\\\").replace(/"/g, "")}"`;

export const assembleFrontmatter = (series: BlogSeriesConfig, today: string, title: string, slug: string, previousPost?: BlogPost): string => {
  const backLink = previousPost ? ` | ${buildBackLink(series, previousPost)}` : "";
  const displayTitle = `${today} | ${series.icon} ${title} ${series.icon}`;
  const quoted = quoteForYaml(displayTitle);
  return `---
share: true
aliases:
  - ${quoted}
title: ${quoted}
URL: ${series.baseUrl}/${today}-${slug}
Author: "${series.author}"
tags:
---
${series.navLink}${backLink}
# ${displayTitle}

`;
};

export const todayPacific = (): string =>
  new Date().toLocaleDateString("en-CA", { timeZone: "America/Los_Angeles" });

File: haskell/src/Automation/BlogPrompt.hs
module Automation.BlogPrompt
  ( BlogContext (..)
  , CrossSeriesPost (..)
  , Slug (..)
  , DisplayTitle (..)
  , mkSlug
  , generateSlug
  , stripEmbedSections
  , buildBlogPrompt
  , filterCommentsAfterLastPost
  , assembleFrontmatter
  , buildDisplayTitle
  , sanitizeTitle
  , recapInstructions
  , buildCrossSeriesSection
  ) where

import Data.Char (isAsciiLower, isDigit)
import Data.List (findIndex)
import Data.Maybe (fromMaybe, mapMaybe)
import Data.Text (Text)
import qualified Data.Text as T
import Data.Time
  ( Day
  , DayOfWeek (..)
  , LocalTime (..)
  , UTCTime
  , addDays
  , dayOfWeek
  , fromGregorian
  , toGregorian
  )
import Data.Time.Format.ISO8601 (iso8601ParseM)
import Text.Read (readMaybe)

import Automation.BlogComments (BlogComment)
import qualified Automation.BlogComments as Comments
import Automation.BlogPosts (BlogPost (..))
import Automation.BlogSeriesConfig (BlogSeriesConfig (..))

import Automation.Frontmatter (quoteYamlValue)
import Automation.PacificTime (formatDay, formatDayHuman, toPacificLocalTime)
import Automation.Text (isEmoji)
import Automation.Wikilink (buildBackLink)
import qualified Automation.Platforms.Bluesky as Bluesky
import qualified Automation.Platforms.Mastodon as Mastodon
import qualified Automation.Platforms.Twitter as Twitter

-- | A validated URL slug: lowercase, alphanumeric + hyphens, no leading/trailing hyphens.
newtype Slug = Slug { unSlug :: Text } deriving (Show, Eq)

-- | A fully constructed display title: "YYYY-MM-DD | icon Title icon".
newtype DisplayTitle = DisplayTitle { unDisplayTitle :: Text } deriving (Show, Eq)

mkSlug :: Text -> Either Text Slug
mkSlug t
  | T.null t = Left "Empty slug"
  | T.any (\c -> c == ' ' || c == '\n') t = Left ("Slug contains whitespace: " <> t)
  | T.head t == '-' || T.last t == '-' = Left ("Slug has leading/trailing hyphens: " <> t)
  | otherwise = Right (Slug t)

-- | A post from another blog series, carrying metadata for prompt formatting.
data CrossSeriesPost = CrossSeriesPost
  { crossSeriesName :: Text
  , crossSeriesIcon :: Text
  , crossSeriesPost :: BlogPost
  } deriving (Show, Eq)

data BlogContext = BlogContext
  { series           :: BlogSeriesConfig
  , agentsMd         :: Text
  , previousPosts    :: [BlogPost]
  , comments         :: [BlogComment]
  , today            :: Day
  , crossSeriesPosts :: [CrossSeriesPost]
  } deriving (Show, Eq)

stripEmbedSections :: Text -> Text
stripEmbedSections content =
  let embedHeaders = [Twitter.sectionHeader, Bluesky.sectionHeader, Mastodon.sectionHeader]
      findPosition header =
        let (before, match) = T.breakOn header content
        in if T.null match then Nothing else Just (T.length before)
      positions = mapMaybe findPosition embedHeaders
  in case positions of
    [] -> T.strip content
    ps -> T.strip (T.take (minimum ps) content)

buildBlogPrompt :: BlogContext -> (Text, Text)
buildBlogPrompt ctx =
  let systemPrompt =
        if T.null (T.strip (agentsMd ctx))
          then defaultSystemPrompt
          else agentsMd ctx
      userPrompt = buildUserPrompt ctx
  in (systemPrompt, userPrompt)

filterCommentsAfterLastPost :: BlogSeriesConfig -> [BlogPost] -> [BlogComment] -> [BlogComment]
filterCommentsAfterLastPost _ [] comments = comments
filterCommentsAfterLastPost series (latestPost : _) comments =
  let postDay = fromMaybe (fromGregorian 2026 1 1) (parseDate (bpDate latestPost))
      cutoff = LocalTime postDay (scheduleTime series)
  in filter (commentAfterCutoff cutoff) comments

commentAfterCutoff :: LocalTime -> BlogComment -> Bool
commentAfterCutoff cutoff comment =
  case parseUtcTimestamp (Comments.createdAt comment) of
    Nothing      -> True
    Just utcTime -> toPacificLocalTime utcTime >= cutoff

parseUtcTimestamp :: Text -> Maybe UTCTime
parseUtcTimestamp = iso8601ParseM . T.unpack

buildDisplayTitle :: BlogSeriesConfig -> Day -> Text -> DisplayTitle
buildDisplayTitle series today title =
  DisplayTitle $ formatDay today <> " | " <> icon series <> " " <> title <> " " <> icon series

sanitizeTitle :: BlogSeriesConfig -> Text -> Text
sanitizeTitle series raw =
  T.strip $ stripTrailingIcon $ stripLeadingIcon $ stripDatePipe $ stripLeadingIcon $ T.strip raw
  where
    stripLeadingIcon t =
      maybe (T.stripStart t) T.stripStart
        (T.stripPrefix (icon series) (T.stripStart t))

    stripTrailingIcon t =
      maybe (T.stripEnd t) T.stripEnd
        (T.stripSuffix (icon series) (T.stripEnd t))

    stripDatePipe t =
      let s = T.stripStart t
      in case parseDate (T.take 10 s) of
        Just _ ->
          let afterDate = T.drop 10 s
              tryStrip = T.stripPrefix " | " afterDate
                     >>= Just . T.stripStart
          in fromMaybe (T.stripStart afterDate) tryStrip
        Nothing -> s

assembleFrontmatter :: BlogSeriesConfig -> Day -> Text -> Slug -> Text
assembleFrontmatter series day title slug =
  let (DisplayTitle displayTitle) = buildDisplayTitle series day title
      url = baseUrl series <> "/" <> formatDay day <> "-" <> unSlug slug
  in T.intercalate "\n"
    [ "---"
    , "share: true"
    , "aliases:"
    , "  - " <> quoteYamlValue displayTitle
    , "title: " <> quoteYamlValue displayTitle
    , "URL: " <> quoteYamlValue url
    , "Author: " <> quoteYamlValue (author series)
    , "---"
    ]

recapInstructions :: Day -> Text
recapInstructions day =
  let dow = dayOfWeek day
      (_, month, dayNum) = toGregorian day
      nextDay = addDays 1 day
      (_, nextMonth, _) = toGregorian nextDay
      isSunday = dow == Sunday
      isLastDayOfMonth = nextMonth /= month
      isLastDayOfQuarter = isLastDayOfMonth && month `elem` [3, 6, 9, 12]
      isLastDayOfYear = month == 12 && dayNum == 31
      instructions = filter (not . T.null)
        [ if isSunday then weeklyRecap else ""
        , if isLastDayOfMonth then monthlyRecap else ""
        , if isLastDayOfQuarter then quarterlyRecap else ""
        , if isLastDayOfYear then yearlyRecap else ""
        ]
  in T.intercalate "\n\n" instructions

defaultSystemPrompt :: Text
defaultSystemPrompt = "You are a creative blog writer. Write engaging, thoughtful blog posts."

buildUserPrompt :: BlogContext -> Text
buildUserPrompt BlogContext{..} =
  let header = "Today is " <> formatDayHuman today <> "."
        <> "\nWrite the next blog post for the " <> name series <> " series."
        <> "\nToday's date: " <> formatDay today
        <> "\n\nIMPORTANT: Your heading (# or ##) must contain ONLY the creative title."
        <> " Do not include dates, pipe separators, or the series icon emoji in your heading."
        <> " The system adds date and icon formatting automatically."
      postHistory = buildPostHistory series previousPosts
      crossSeriesContext = buildCrossSeriesSection crossSeriesPosts
      commentsSection = buildCommentsSection comments
      recap = recapInstructions today
  in T.intercalate "\n\n"
    $ filter (not . T.null)
      [ header
      , postHistory
      , crossSeriesContext
      , commentsSection
      , recap
      ]

buildPostHistory :: BlogSeriesConfig -> [BlogPost] -> Text
buildPostHistory _ [] = ""
buildPostHistory series posts =
  let recentPosts = postsSinceLastRecap posts
      formatted = fmap (formatPost series) recentPosts
  in "## Recent Posts\n\n" <> T.intercalate "\n\n---\n\n" formatted

postsSinceLastRecap :: [BlogPost] -> [BlogPost]
postsSinceLastRecap posts =
  let capped = take 7 posts
  in case findIndex isRecapPost capped of
    Nothing -> capped
    Just i  -> take (i + 1) capped

isRecapPost :: BlogPost -> Bool
isRecapPost = T.isInfixOf "recap" . T.toLower . bpTitle

formatPost :: BlogSeriesConfig -> BlogPost -> Text
formatPost series post =
  let link = buildBackLink series (bpFilename post)
      body = stripEmbedSections (bpBody post)
      cleanTitle = sanitizeTitle series (bpTitle post)
  in "### " <> link <> " " <> cleanTitle <> " (" <> bpDate post <> ")\n\n" <> T.strip body

buildCommentsSection :: [BlogComment] -> Text
buildCommentsSection [] = ""
buildCommentsSection comments =
  let formatted = fmap formatComment comments
  in "## Reader Comments\n\n" <> T.intercalate "\n\n" formatted

buildCrossSeriesSection :: [CrossSeriesPost] -> Text
buildCrossSeriesSection [] = ""
buildCrossSeriesSection posts =
  let formatted = fmap formatCrossSeriesPost posts
  in "## Today Across the Blog\n\n"
    <> "The following are the most recent posts from other blog series on this site. "
    <> "Each series has its own voice and perspective. Find connections, tensions, "
    <> "and emergent themes across these independent voices.\n\n"
    <> T.intercalate "\n\n---\n\n" formatted

formatCrossSeriesPost :: CrossSeriesPost -> Text
formatCrossSeriesPost CrossSeriesPost{..} =
  let body = stripEmbedSections (bpBody crossSeriesPost)
      excerpt = T.take 2000 body
  in "### " <> crossSeriesIcon <> " " <> crossSeriesName <> " — " <> bpTitle crossSeriesPost
    <> " (" <> bpDate crossSeriesPost <> ")\n\n" <> T.strip excerpt

formatComment :: BlogComment -> Text
formatComment c =
  let priority = if Comments.isPriority c then " ⭐" else ""
  in "**" <> Comments.author c <> "**" <> priority <> " (" <> Comments.createdAt c <> "):\n" <> Comments.body c

parseDate :: Text -> Maybe Day
parseDate t =
  case T.splitOn "-" t of
    [yStr, mStr, dStr] ->
      fromGregorian
        <$> readMaybe (T.unpack yStr)
        <*> readMaybe (T.unpack mStr)
        <*> readMaybe (T.unpack dStr)
    _ -> Nothing

weeklyRecap :: Text
weeklyRecap =
  "📅 Today is Sunday. Please include a brief weekly recap section "
    <> "summarizing the key themes and developments from this week's posts."

monthlyRecap :: Text
monthlyRecap =
  "📆 Today is the last day of the month. Please include a monthly recap "
    <> "section summarizing the key themes and developments from this month."

quarterlyRecap :: Text
quarterlyRecap =
  "📊 Today is the last day of the quarter. Please include a quarterly recap "
    <> "section summarizing the key themes from the past three months."

yearlyRecap :: Text
yearlyRecap =
  "🎆 Today is the last day of the year. Please include a yearly recap "
    <> "section reflecting on the major themes and developments from the entire year."

generateSlug :: Text -> Text
generateSlug title =
  let cleaned = T.filter (not . isEmoji) title
      lowered = T.toLower (T.strip cleaned)
      alphanum = T.map (\c -> if isAlphaNumOrSpace c then c else ' ') lowered
      dashed = T.intercalate "-" (T.words alphanum)
      trimmed = T.dropWhile (== '-') (T.dropWhileEnd (== '-') dashed)
  in trimmed
  where
    isAlphaNumOrSpace c = isAsciiLower c || isDigit c || c == ' ' || c == '-'
